### PR TITLE
implement remove and rename feature for fat32

### LINKED ISSUE #191: Support to remove files
Hi, thanks for the awesome library.

I played around a bit with the library and was able to create new files and upate existing files on a fat32 partition inside an image. Unfortunately, there seems to be no support for removing files. Do I miss something? If its really missing, will the remove functionality come in the future?

Thanks and best regards

### PR DESCRIPTION
**Pull request by ludelion5**
Sorry for the late PR. Here is my proposal to rename and remove files in fat32.

closes #191

### CODE REVIEW

**Review by deitch — COMMENTED** [review]
Overall, this looks great, and thank you.

In terms of naming, are these meant to be restricted just to _files_, or any structure (files, dirs, symlinks where supported, etc.)? I would think the latter, since CLI-level commands for the above in most OSes don't care. Same for linux `mv` / Windows `ren`.

So should these be `RemoveFile()` and `RenameFile()`, or just `Remove()` and `Rename()`?

On the other hand, in keeping with golang style, maybe these should parallel and be [`Remove()`](https://pkg.go.dev/os#Remove) and [`RemoveAll()`](https://pkg.go.dev/os#RemoveAll) and [`Rename()`](https://pkg.go.dev/os#Rename]?

**Comment by ludelion5** [comment]
Thanks for the feedback. I will look into your request and test, if the implementation will also work for symlinks and directories. If so, its a good idea to stick to your suggested naming-scheme.

Additionally I can check if the RemoveAll is easy to implement, I have to see if I can spare enough time for that feature.

**Comment by ludelion5** [comment]
Sorry for the late update on this PR.

We adapted our Rename and Remove methods to the new interface.

Additionally we have implemented the ability to remove empty directories and the renaming to already existing files (overwriting).
Currently, the renaming (move) works only in the same directory, this would require a deeper understanding of your `fs.allocateSpace`, I guess this method has to be called multiple times, depending on how many directory levels the file moves. We check if the base-dir is different and return an error.

File: filesystem/fat32/fat32.go [hunk 1]
```diff
@@ -637,18 +637,130 @@ func (fs *FileSystem) OpenFile(p string, flag int) (filesystem.File, error) {
...
-// Rename renames (moves) oldpath to newpath. If newpath already exists and is not a directory, Rename replaces it.
-//
-//nolint:revive // parameters will be used eventually
-func (fs *FileSystem) Rename(oldpath, newpath string) error {
-	return filesystem.ErrNotImplemented
-}
-
 // removes the named file or (empty) directory.
 //
 //nolint:revive // parameters will be used eventually
```
Comment (aol-nnov) [inline filesystem/fat32/fat32.go:642]: As the method got implemented, lint exception is not needed anymore here. Please remove it.

**Comment by deitch** [comment]
> Currently, the renaming (move) works only in the same directory, this would require a deeper understanding of your fs.allocateSpace, I guess this method has to be called multiple times, depending on how many directory levels the file moves. We check if the base-dir is different and return an error

That certainly is enough for now. We go from "this is not implemented" to "this is partially implemented", which his an improvement. So thank you.

**Review by deitch — CHANGES_REQUESTED** [review]
Just the lint comment.

**Comment by ludelion5** [comment]
done

**Review by deitch — APPROVED** [review]

**Comment by deitch** [comment]
Just some lint errors that need to be fixed.

**Comment by ludelion5** [comment]
`make lint` should now be satisfied

**Review by deitch — APPROVED** [review]

**Comment by deitch** [comment]
Looks great, thanks!

**Comment by ludelion5** [comment]
are you planning to create a new tag as well?

**Comment by deitch** [comment]
Yeah but let's wait for this one.

**Comment by ludelion5** [comment]
Do you have an estimation when a new tag is created?

**Comment by deitch** [comment]
https://github.com/diskfs/go-diskfs/releases/tag/v1.5.0

**Pull request merged**